modcc/solvers.hpp

```diff
@@ -83,6 +82,28 @@ class SparseSolverVisitor : public SolverVisitorBase {
     // 'Symbol table' for symbolic manipulation.
     symge::symbol_table symtbl_;
 
+public:
+    using SolverVisitorBase::visit;
+
+    explicit MatrixSolverVisitor() {}
+    MatrixSolverVisitor(scope_ptr enclosing): SolverVisitorBase(enclosing) {}
+
+    virtual void reset() override {
+        deq_index_ = 0;
+        local_expr_.clear();
+        A_.clear();
+        symtbl_.clear();
+        SolverVisitorBase::reset();
+    }
+
+    virtual std::vector<expression_ptr> GenerateReduceStatements();
+    virtual std::vector<expression_ptr> GenerateUpdateStatements(std::vector<std::string>);
+};
+
```

Comment:
I think this would be better implemented as a class outside of the Visitor hierarchy: with the exception of `reset()`, it's not providing any implementations of Visitor or SolverVisitorBase methods, and neither are the `Generate...()` overridden by its derived classes.

It could be implemented as a separate class, used in the solvers as a member, or as a mix-in, which is inherited by the solvers.

In terms of isolation or interface, there a few issues:
* `deq_index_` and `local_expr_` are only used in the solvers, and so shouldn't be in this class.
* By convention in the modcc code, method names are given in lower_case_with_underscores; this convention should be adopted for `GenerateReduceStatements()` etc.
* `GenerateReduceStatements()` both writes to `statements_` _and_ returns a list of expressions to be used by the caller — this makes for a more complex and difficult to understand interface. Given that the temporaries have to be declared as local once, but assigned to multiple times in the nonlinear solver, it is probably best to have a method for generating the local declarations (without assignments), and another which returns the list of statements for the reduction.
*  `GenerateUpdateStatements()` — I'm being a bit picky here, but it's not really clear to me from the name of the function or the name of the parameter what it actually does. How about something like `generate_solution_assignments(std::vector<std::string> lhs_vars)`?

In terms of the implementation (and I didn't see this before), aren't the rescalings happening too late? The temporary fill-ins will grow in magnitude through the series of assignments generated from the symbol table (`modcc/solvers.cpp`:185). It may still be that the easiest way to get the scaling happening at the right place is to extend the functionality of symge, allowing symbols to be bound to `symbol_term_diff` expressions _or_ a new symbol reciprocal expression _or_ explicitly 1. Before pivoting, we could optionally scale the row: make a new sym for the reciprocal of the pivot element, and replace the row with terms that are the product of the old elements and this reciprocal.